### PR TITLE
Support fused MLA QKV checkpoint reload

### PR DESCRIPTION
**Pull request by sraman-rgb**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/core/optimizer/distrib_optimizer.py [hunk 1]
```diff
@@ -2882,6 +2882,34 @@ def _normalize_state_dict_for_grouped_params(state_dict_flat, model_chunk):
...
+    @staticmethod
+    def _synthesize_state_dict_params_for_model(state_dict_flat, model_chunk):
+        """Let modules materialize runtime params before optimizer state-dict matching.
+
+        This covers modules whose runtime parameter is assembled from multiple checkpoint
+        tensors. Normal model loading can handle this in _load_from_state_dict(), but
+        reload_model_params(state_dict=...) matches optimizer master params directly by name.
+        """
+        for module_name, module in model_chunk.named_modules():
+            synthesize = getattr(module, '_synthesize_fused_qkv_down_weight', None)
```
Comment (santhnm2) [inline megatron/core/optimizer/distrib_optimizer.py:2896]: Can you add a comment about what this `synthesize` function does / what arguments it takes?

File: megatron/core/optimizer/distrib_optimizer.py [hunk 2]
```diff
@@ -2882,6 +2882,34 @@ def _normalize_state_dict_for_grouped_params(state_dict_flat, model_chunk):
...
+            if not callable(synthesize):
+                continue
+
+            clean_name = module_name
+            while clean_name.startswith("module."):
+                clean_name = clean_name[len("module.") :]
+
+            module_prefix = f"{clean_name}." if clean_name else ""
```
Comment (santhnm2) [inline megatron/core/optimizer/distrib_optimizer.py]: Is there an existing utility that can extract the prefix? This routine seems a bit brittle / not very future-proof in case the module name format changes at some point

File: megatron/core/optimizer/distrib_optimizer.py [hunk 3]
```diff
@@ -2882,6 +2882,34 @@ def _normalize_state_dict_for_grouped_params(state_dict_flat, model_chunk):
...
+            synthesize(state_dict_flat, module_prefix)
+
+            q_weight_suffix = f"{module_prefix}linear_q_down_proj.weight"
```
Comment (santhnm2) [inline megatron/core/optimizer/distrib_optimizer.py]: I'm assuming this `linear_q_down_proj` string is safe to hard-code here?

File: tests/unit_tests/test_optimizer.py [hunk 4]
```diff
@@ -806,6 +807,33 @@ def test_optimizer_reload_model_params():
...
+    model.decoder.layers = nn.ModuleList([nn.Module()])
+    model.decoder.layers[0].self_attention = FakeFusedMLA()
+
+    prefix = "module.decoder.layers.0.self_attention."
+    q_key = f"{prefix}linear_q_down_proj.weight"
+    kv_key = f"{prefix}linear_kv_down_proj.weight"
+    fused_key = f"{prefix}linear_qkv_down_proj.weight"
+    q_weight = torch.ones(2, 3)
+    kv_weight = torch.full((4, 3), 2.0)
+    state_dict = {q_key: q_weight, kv_key: kv_weight}
```
Comment (santhnm2) [inline tests/unit_tests/test_optimizer.py]: Faking the state dict keys here seems brittle since these may diverge from the real layer's state dict keys, can we instead instantiate a real layer here and directly use its `state_dict`?

File: megatron/core/optimizer/distrib_optimizer.py (hunk 1 above)
Comment (sraman-rgb) [reply megatron/core/optimizer/distrib_optimizer.py:2896]: addressed in the latest commit

File: megatron/core/optimizer/distrib_optimizer.py (hunk 3 above)
Comment (sraman-rgb) [reply megatron/core/optimizer/distrib_optimizer.py]: addressed in the latest commit

File: tests/unit_tests/test_optimizer.py (hunk 4 above)
Comment (sraman-rgb) [reply tests/unit_tests/test_optimizer.py]: addressed in the latest commit

File: megatron/core/optimizer/distrib_optimizer.py (hunk 2 above)
Comment (sraman-rgb) [reply megatron/core/optimizer/distrib_optimizer.py]: addressed in the latest commit

**Review by WanZzzzzz — COMMENTED** [review]
LGTM.

**Review by WanZzzzzz — APPROVED** [review]

File: megatron/core/transformer/multi_latent_attention.py [hunk 5]
```diff
@@ -1429,8 +1432,12 @@ def _clone_sharded_object_with_key(obj: ShardedObject, new_key: str) -> ShardedO
 
         return sharded_state_dict
 
-    def _load_from_state_dict(self, state_dict, prefix, *args, **kwargs):
-        """Load state dict with automatic unfused->fused conversion."""
+    def _synthetic_state_dict_key_suffixes(self):
+        """Return source checkpoint keys used to locate this module in a state dict."""
+        return ("linear_q_down_proj.weight",)
```
Comment (santhnm2) [inline megatron/core/transformer/multi_latent_attention.py:1437]: Is this expected to grow to more than 1 value? If so it makes sense to keep this as a function but otherwise the function seems like an unnecessary abstraction.

**Review by santhnm2 — APPROVED** [review]

File: megatron/core/transformer/multi_latent_attention.py (hunk 5 above)
Comment (sraman-rgb) [reply megatron/core/transformer/multi_latent_attention.py:1437]: yes, there is another thing that i am working on, which will need change here (unless we do something differently)

**Review by jaredcasper — APPROVED** [review]
